google/cloud/storage/client.cc

```diff
@@ -233,6 +233,41 @@ credentials, as described in:
   return std::move(os).str();
 }
 
+StatusOr<std::string> Client::SignUrlV4(internal::V4SignUrlRequest request) {
+  request.AddMissingRequiredHeaders();
+  auto base_credentials = raw_client()->client_options().credentials();
+  auto credentials = dynamic_cast<oauth2::ServiceAccountCredentials<>*>(
+      base_credentials.get());
+
+  if (credentials == nullptr) {
+    return Status(StatusCode::kInvalidArgument,
+                  R"""(The current credentials cannot be used to sign URLs.
+Please configure your google::cloud::storage::Client to use service account
+credentials, as described in:
+https://cloud.google.com/storage/docs/authentication
+)""");
+  }
+
+  auto string_to_sign = request.StringToSign(credentials->client_id());
+  auto result = credentials->SignStringHex(string_to_sign);
+  if (!result.first.ok()) {
+    return result.first;
+  }
+
+  auto signature = result.second;
+
+  internal::CurlHandle curl;
+  std::ostringstream os;
+  os << "https://storage.googleapis.com/" << request.bucket_name();
+  if (!request.object_name().empty()) {
+    os << '/' << curl.MakeEscapedString(request.object_name()).get();
+  }
+  os << "?" << request.CanonicalQueryString(credentials->client_id())
+     << "&X-Goog-Signature=" << signature;
+
+  return std::move(os).str();
```

Comment:
I remember hearing Titus say that future versions of the standard would do the right thing here (move the contents instead of copying), but I cannot find a reference.  Removed.

Reply:
[p0408r6](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0408r6.pdf) is what you're looking for I think.

Reply:
I'm not sure it is worth coding to possible future versions of the standard library, but there are similar snippets (e.g., on line 233) so it is a bit of a trade off between consistency and clarity.  Your choice.

Reply:
I will go with `std::move(os)` for now.  If it becomes an annoyance we can change it everywhere. Also (and this might be purely rationalization), our *customers* may be compiling with C++17 (and C++20), even if we are not.